src/main/scala/mesosphere/marathon/api/v2/TasksResource.scala

```diff
@@ -30,22 +30,44 @@ class TasksResource @Inject() (
   @GET
   @Produces(Array(MediaType.APPLICATION_JSON))
   @Timed
-  def indexJson(@QueryParam("status") status: String,
-                @QueryParam("status[]") statuses: util.List[String]): Response = {
+  def indexJson(
+    @QueryParam("status") status: String,
+    @QueryParam("status[]") statuses: util.List[String],
+    @QueryParam("limit")@DefaultValue("-1") limit: Int,
+    @QueryParam("offset")@DefaultValue("0") offset: Int): Response = {
     if (status != null) statuses.add(status)
     val statusSet = statuses.asScala.flatMap(toTaskState).toSet
+
+    val tasks = taskTracker.list.values.iterator.flatMap { app =>
+      app.tasks.iterator.map(t => app.appName -> t)
+    }
+
+    val appToPorts = taskTracker.list.keySet.map { appId =>
+      appId -> service.getApp(appId).map(_.servicePorts).getOrElse(Nil)
```

Comment:
I would probably collect the appIds I need first and then only query them for appToPorts and health. That especially makes sense if you integrate paging above.